Add countdown screen to tournament client

CloneWith opened the pull request:
- [x] Core countdown design
- [x] Upcoming match section
- [x] Schedule section
- [x] ~Auto progress between states (countdown only -> +-schedule -> +-upcoming)~ (**Aborted**)
- [x] L10N

CloneWith committed:
Add triangles to countdown background

CloneWith committed:
Increase animation speed when time is not enough

To be precise, it means no more than one minute...

CloneWith committed:
Remake the timer complete animation

CloneWith committed:
Add first select trigger support for TournamentScreen

CloneWith committed:
Fill the full placeholder content when it's first loaded

CloneWith committed:
Add countdown screen (stub)

CloneWith committed:
Temporarily fix `zh` CultureInfo unable to humanize date offsets

CloneWith committed:
Add upcoming match section to CountdownScreen

NaughtyChas commented:
Don't forget we have to add the logo display as well.

- [x] ~Logo display (including variants)~ (**Transferred, see #70**)

also the decoration sprite on the countdown widget...

- [x] Custom deco sprite above the countdown widget

CloneWith committed:
Add recent matches schedule section to CountdownScreen

NaughtyChas commented:
> - [x] ~Auto progress between states (countdown only -> +-schedule -> +-upcoming)~ (**Aborted**)

In my initial thoughts, these three states are not sequential but parallel. This means there is no progression among them. The goal is to provide our streamers with three different options for displaying the countdown in their preferred manner. Streamers can choose to display only the upcoming events, the full schedule, or a simpler countdown only.

Also removing this will save us some work so it might be better for us :)

CloneWith committed:
Localize countdown strings

NaughtyChas committed:
Use container for Hourglass icon

So when it spins it won't affect the position of the text

NaughtyChas committed:
Finetune Hourglass animation

CloneWith committed:
Add iconContainer instead of indicatorIcon to contentFlow

NaughtyChas committed:
Improve countdown intro animation

Imma param tweaker lol

NaughtyChas reviewed (commented):
After finetuning some animation params everything looking good.

btw have you tried passing through `Set a target time in the past` step? Looks like there is an error caused by the `getWaitingString()` method trying to format a negative time interval when the countdown target time is set to a past time.

There will be an uncaught exception when the countdown ends (at 00:00.000). I still haven't found a stable way to recreate this, so I will update this comment as soon as there is progress. Better to keep an eye on this one.

CloneWith committed:
Remove duplicate gitignore entry

CloneWith commented:
> After finetuning some animation params everything looking good.
> 
> btw have you tried passing through `Set a target time in the past` step? Looks like there is an error caused by the `getWaitingString()` method trying to format a negative time interval when the countdown target time is set to a past time.
> 
> There will be an uncaught exception when the countdown ends (at 00:00.000). I still haven't found a stable way to recreate this, so I will update this comment as soon as there is progress. Better to keep an eye on this one.

This exception was always raisen when debugging, and it was found out to be caused by malformed string interpolating. Gotta fix this.

CloneWith committed:
Use the right format for string interpolating

Should always be {0} instead of {}

CloneWith committed:
Add kanban element sprite for CountdownScreen

CloneWith committed:
Add subtle scaling to kanban's animation

CloneWith committed:
Add looped floating animation to kanban element

CloneWith commented:
> Don't forget we have to add the logo display as well.
> 
> - [x] Logo display (including variants)

This can be extracted from this issue and into an individual one since other screens might make use of that in the future.

NaughtyChas committed:
Add subtle rotation to sprite

NaughtyChas committed:
Add hourglass rotation back

turns out the issue does not came from negative param of Then, but I accidentally removed this part of code

NaughtyChas commented:
> > Don't forget we have to add the logo display as well.
> > 
> > * [x]  Logo display (including variants)
> 
> This can be extracted from this issue and into an individual one since other screens might make use of that in the future.

I am thinking if we can use the existing [`OnlineMenuBanner.cs`](https://github.com/CloneWith/osu/blob/derived/new/osu.Game/Screens/Menu/OnlineMenuBanner.cs) to achieve this effect. Just change the image source then we're good.

Tracked using issue #70 .

NaughtyChas committed:
Code formatting

CloneWith commented:
> > > Don't forget we have to add the logo display as well.
> > > 
> > > * [x]   Logo display (including variants)
> > 
> > 
> > This can be extracted from this issue and into an individual one since other screens might make use of that in the future.
> 
> I am thinking if we can use the existing [`OnlineMenuBanner.cs`](https://github.com/CloneWith/osu/blob/derived/new/osu.Game/Screens/Menu/OnlineMenuBanner.cs) to achieve this effect. Just change the image source then we're good.

Sounds good, would have a try.

CloneWith commented:
A remaining problem has been found during tests, would be fixed soon and it is good to merge after this:

- [x] Improper state when changing from `Ended` to `In progress` or `Unset`

CloneWith committed:
Reset the countdown when an ending animation has been triggered but not handled

NaughtyChas reviewed (commented):
Bro is writing robust code, and I am struggling to find a bug in it. 

As for the logo display should we track it in a new PR, or finish it in this PR instead?

CloneWith commented:
> Bro is writing robust code, and I am struggling to find a bug in it.
> 
> As for the logo display should we track it in a new PR, or finish it in this PR instead?

I prefer tracking it in a new PR, since we still need to consider its design and other details, and it could be used in other places.
